examples/gno.land/r/gnoland/events/events.gno

```diff
@@ -1,240 +1,316 @@
 
-<div class="column">
+	std.Emit(DeletionEvent,
+		"id",
+		e.id.String(),
+	)
+}
 
-### GopherCon EU
+// EditEvent edits an event with a given ID
+// It only updates values corresponding to non-empty/non-nil arguments sent with the call
+func EditEvent(id seqid.ID, name, description, link, location string, startTime, endTime int64) {
+	a.AssertOnAuthList()
 
-- **Come Meet Us at our Booth**
-- Berlin, July 26 - 29, 2023
+	e, _, err := GetEventByID(id)
+	if err != nil {
+		panic(err)
+	}
 
-[Learn more](https://gophercon.eu/)
+	// Set only valid values
+	if name != "" {
+		e.name = name
+	}
 
-</div><!-- end column-->
+	if description != "" {
+		e.description = description
+	}
 
-<div class="column">
+	if link != "" {
+		e.link = link
+	}
 
-### Nebular Summit Gno.land for Developers
+	if location != "" {
+		e.location = location
+	}
 
-- Paris, July 24 - 25, 2023
-- Manfred Touron
+	if startTime > 0 {
+		e.startTime = time.Unix(startTime, 0)
+	}
 
-[Learn more](https://www.nebular.builders/)
+	if endTime > 0 && endTime > startTime {
+		e.endTime = time.Unix(endTime, 0)
+	} else {
+		panic(ErrInvalidEndTime)
+	}
 
-</div><!-- end column-->
+	std.Emit(EditingEvent,
+		"id",
+		e.id.String(),
+	)
 
-<div class="column">
+	// If sort order was disrupted, sort again
+	if startTime > 0 {
+		sort.Sort(events)
+	}
+}
 
-### EthCC
+// Rendering
 
-- **Come Meet Us at our Booth**
-- Paris, July 17 - 20, 2023
-- Manfred Touron
+// RenderEventWidget shows up to amt of the latest events to a caller
+func RenderEventWidget(amt int) string {
+	eventNum := len(events)
+	if eventNum == 0 {
+		return "No events."
+	}
 
-[Learn more](https://www.ethcc.io/)
+	if amt > 5 {
+		panic("maximum number of events in widget is " + strconv.Itoa(maxWidgetSize))
+	}
 
-</div><!-- end column-->
+	if amt < 1 {
+		panic("amt needs to be more than 0")
+	}
 
-<div class="column">
+	if amt > eventNum {
+		amt = eventNum
+	}
 
-### Eth Seoul
+	output := ""
 
-- **The Evolution of Smart Contracts: A Journey into Gno.land**
-- Seoul, June 3, 2023
-- Manfred Touron
+	for _, event := range events[eventNum-amt:] {
+		if event.startTime.After(time.Now()) {
+			output += ufmt.Sprintf("- [%s](%s)\n", event.name, event.link)
+		}
+	}
 
-[Learn more](https://2023.ethseoul.org/)
+	return output
+}
 
-</div><!-- end column-->
-<div class="column">
+func renderHome(admin bool) string {
+	output := "# gno.land events\n\n"
 
-### BUIDL Asia
+	if len(events) == 0 {
+		output += "No upcoming or past events."
+		return output
+	}
 
-- **Proof of Contribution in Gno.land**
-- Seoul, June 6, 2023
-- Manfred Touron
+	output += "Below is a list of all upcoming and past gno.land events.\n\n"
+	output += ufmt.Sprintf(time.Now().Format(time.RFC850))
 
-[Learn more](https://www.buidl.asia/)
+	output += "---\n\n"
 
-</div><!-- end column-->
-<div class="column">
+	// Find position of la1test past event
+	pos := sort.Search(len(events), func(i int) bool {
+		return events[i].startTime.Before(time.Now())
```

Comment:
Actually this won't work. If there is an event with a long duration it is possible that it ends during another in progress event,  but the events are sorted by start time, so it could incorrectly mark an in progress event as complete.

Current time: 5 PM
Event[0] -- start 12PM, end 4PM
Event[1] -- start 1PM, end 8PM

The index of ended events would be 0 but 1 is still going on. We could think about how to better support in progress events. Any ideas?

Reply:
Per our discussion, this probably isn't necessary.